Adds opium and meth rocks!

massaheartsu opened the pull request:
## About The Pull Request

![demo2](https://github.com/user-attachments/assets/26223fa6-6eb8-43eb-ba87-54be7d2c19b3)<img width="612" height="185" alt="dreamseeker_SNMvxqCRiR" src="https://github.com/user-attachments/assets/1d3103e9-edc1-4e50-8793-c8c2ace53aea" />

This is a small expansion to narcotics, their production, and presence in the game! For my esteemed Terry and Manuel/Sybil sector colleagues I've added:
- 2 new solid drug items: **crystal meth and opium**!
- A **precipitation reaction** for meth after it's been cooled, producing a solid crystal form, 2u sulfuric acid, 10u meth. **Above 90% purity, the meth will gain an increasingly blue tint!**
- A production method for opium! Slice a well developed poppy pod with any sharp item, **before it's fully mature and flowered** to extract a small amount of opium to enjoy! It's a small container for morphine. Can be pressed together to combine the concentrations up to 10u, as each extraction contains trace amounts. Scales off poppy's potency.
- Populated narcotics/contraband spawners with new items, and items that really should have been on there.
- Added a brand new smoking apparatus, the glass pipe! Can be crafted. Fixed the transparent pixel.
- **Fixed methsplosions so that now spacemen can finally smoke meth without exploding!!!** Omegaweed too. Normal methsplosions are untouched!
- **Expanded the ability for players to insert all drug items into pipes**, so now players can smoke moon rocks, SaturnX, and my new items! Previously, only 'dried' items were.
- Fentanyl patch box for narcotics spawner, also for a new ruin I've made.
- **Expanded these changes to the black market uplink.** 
<img width="640" height="256" alt="demo" src="https://github.com/user-attachments/assets/c8a0eb2c-b0fa-4e70-b6c1-2e741cd170a2" />

Now, players can precipitate meth into a solid form, and produce opium. I added a lot of these substances and the ones that didn't exist into spawners that they were partially in, or could reasonably be in. The opium poppy extraction required a new variable that can be used for similar interactions too. Crystal sprites are transparent, smoked sprite has a nice little animation and glow. 

These items are now on the black market uplink, in both expensive dealer shipments of varying rarities for different narcotics, as well as personal use quantities on a different rotation, at a steeper price.  Expanded randomized spawners to give the black market and ruins more teeth and edge, without adding another weapon.

Credit to the moon rock guy, I'm just happy I made your stuff smokable!!
<img width="219" height="201" alt="dreamseeker_N8x9bsN3kS" src="https://github.com/user-attachments/assets/64d23c27-92b9-423b-91ac-903b8476d0b4" />


tl;dr added opium and meth
## Why It's Good For The Game
For **too long**, narcotics, chems and stims have been mere integers or buffs/debuffs in pills or syringes, so I introduced two new methods for making them and populated niche, criminal vectors with them. These are real items now, with more interactions. Allowing these to be smoked also allowed moon rocks and SaturnX to be smoked too, including them into this and enticing players to have another reason to make them as well. Buffs with drawbacks, addictions and problems are actually great. 

It's a new venue for flavor. I've put thought into the balance for addition and quantity. It'll add a criminal, or exotic tone to a round, provide new business opportunities, and new access to stims and chems without going over the top. Perhaps something peaceful antagonists can do. More proverbial floor pills give security or command something to selectively enforce, and absolutely provide novel opportunities for medical staff!

Also, methsplosions now don't happen in cigarette items! You can smoke omegaweed without fucking dying!
## Changelog
:cl:
add: Added meth, opium, and their production! Also added to black market uplink.
add: Added a craftable glass pipe.
add: Added to contraband spawners. 
qol: Can now insert any drug item into a pipe.
fix: Spacemen can now smoke meth without exploding!!
/:cl:

massaheartsu committed:
first commit

Added new drug items and populated loot lists with them. More TBD.

massaheartsu committed:
added rocks, sap, icons, and code

spent all fucking day arguing with punctuation.

added sap extraction, rock synthesis, code for it, sprites, tested thoroughly.

massaheartsu committed:
working drop

Updated sprites, added functionality, fixed problems. Works, looks good, testing from here.

massaheartsu committed:
Update pyrotechnics.dm

The vital changes necessary for spacemen to smoke meth, while still allowing it to explode.

massaheartsu committed:
nothingburger

massaheartsu committed:
IT WORKS FUCK YES

You can now smoke crystal meth without it blowing all your limbs off!

massaheartsu committed:
Complete!

All functioning!
- Added, tweaked all sprites, tweaked vars and numbers for realistic gameplay.
- Fixed errors.
- Finished code for items, extraction, combining, synthesis, etc. processes.
- Tested thoroughly that everything works, sprites function and are right.
- Allowed spacemen to smoke meth without exploding!
- Added new crafting code.

massaheartsu committed:
Actually complete.

Minor adjustments. Tweaked things, tested, untweaked, fixed descriptions and sprites. Fixed buges.

massaheartsu committed:
made animated pipe transparent

computers are bad, everyone who made them is bad

massaheartsu committed:
actually done

Christ. Added all the items+some fun wholesale packages to the black market. Tweaked, fixed, etc. new box icon.

massaheartsu committed:
dd

changed directory lol

massaheartsu committed:
reduced fent quantity

so many changes

massaheartsu committed:
oops

massaheartsu committed:
Merge branch 'master' into narcontent

necromanceranne commented on code/game/objects/items/cigarettes.dm:
```diff
@@ -939,14 +941,14 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 		return ..()
 
-	var/obj/item/food/grown/to_smoke = thing
+	var/atom/to_smoke = thing
```
Since we're no longer looking for a specific type, we don't actually need to redefine ``thing``, you can just delete this rather than redefining it to its own parent type (which is usually not a good idea in the first place)

massaheartsu replied:
Cleaned this up and made sure it works!

massaheartsu committed:
Added and fixed requested changes!

Adjustments requested by maintainers.

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -19,6 +19,18 @@
 	mutatelist = list(/obj/item/seeds/poppy/geranium, /obj/item/seeds/poppy/lily)
 	reagents_add = list(/datum/reagent/medicine/c2/libital = 0.2, /datum/reagent/consumable/nutriment = 0.05)
 
+/obj/item/seeds/poppy/attackby(obj/item/I, mob/user, obj/machinery/hydroponics/tray) //Allows for extraction of opium sap before full maturation.
```
Done!

massaheartsu commented on code/modules/hydroponics/hydroponics.dm:
```diff
@@ -938,6 +938,12 @@
 
+	else if(O.sharpness) // Allows for the extraction (for opium or sap) interaction if a seed has it.
+		if(myseed && !myseed.extracted)
+			myseed.attackby(O, user, src)
```
Also cleaned this up!

massaheartsu committed:
fixed inhand_icons

oop

massaheartsu committed:
fixed icon states again. also tweaked numbers in hindsight.

thought it'd be a bit better this way. also fixed a typo.

SyncIt21 commented on code/game/objects/items/cigarettes.dm:
```diff
@@ -435,6 +435,8 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 		return
 	// allowing reagents to react after being lit
 	reagents.flags &= ~(NO_REACT)
+	if(reagents && reagents.has_reagent(/datum/reagent/drug/methamphetamine)) //Allows spacemen to smoke meth without exploding!
+		reagents.flags |= NO_REACT
```
```suggestion
	if(!reagents.has_reagent(/datum/reagent/drug/methamphetamine)) //Allows spacemen to smoke meth without exploding!
		reagents.flags &= ~(NO_REACT)
```

SyncIt21 commented on code/game/objects/items/drug_items.dm:
```diff
@@ -81,3 +81,85 @@
+	embed_chance = 7
+	fall_chance = 2
+	rip_time = 2 SECONDS
+	var/transfer_per_second = 0.75
```
This var is unused

massaheartsu replied:
Is it? I took it from syringes, and the meth embeds. It seemed to work. I removed it for now.

SyncIt21 commented on code/game/objects/items/drug_items.dm:
```diff
@@ -81,3 +81,85 @@
+	amount = rand(1, 7) / 10
+	reagents.add_reagent(/datum/reagent/consumable/sugar, amount)
+
+/obj/item/food/drug/opium/raw/attackby(obj/item/I, mob/user) //Allows for combining opium up to 7u of morphine in a ball of opium, leaving the quantity vague-ish until maxed. Any further refinement requires machinery.
```
convert to `interact_with_atom()`

SyncIt21 commented on code/modules/hydroponics/grown/flowers.dm:
```diff
@@ -19,6 +19,18 @@
+		user.visible_message(
+			span_notice("You carefully slice the poppy's pod, collecting the fragrant, alluring sap.")
+		)
+		return
+	..()
```
```suggestion
		return ITEM_INTERACT_SUCCESS
	return NONE
```

SyncIt21 commented on code/modules/reagents/chemistry/recipes/drugs.dm:
```diff
@@ -77,6 +77,20 @@
+
+/datum/chemical_reaction/meth_crystal/on_reaction(datum/reagents/holder, datum/equilibrium/reaction, created_volume)
+	var/location = get_turf(holder.my_atom)
+	for(var/i in 1 to created_volume)
```
```suggestion
	for(var/i in 1 to round(created_volume, CHEMICAL_VOLUME_ROUNDING))
```

Wallemations reviewed (approved):
Anyways sprites are fine I guess, I don't know anything about real drugs so I can't really use any sort of frames of reference

mc-oofert commented:
going to Google "how to make crystal meth ss13" but I forget to type ss13 and I get put on a watchlist

massaheartsu committed:
Update code/modules/hydroponics/grown/flowers.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

massaheartsu committed:
Update code/modules/reagents/chemistry/recipes/drugs.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

massaheartsu committed:
Update code/game/objects/items/cigarettes.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

massaheartsu committed:
fixed another attackby

fixed another attackby

massaheartsu commented:
> I'll be honest, I think it would be better if we came up with fake drugs rather than using real ones.
> 
> Feels weird to me to use more real ones than the ones we already got, but that might just be me.

I actually felt quite the opposite! I wanted what we have right now to not just be variables in pills or syringes. We have kronkaine and saturnX, which I opened the door to, I wanted to open just a small new vector. But, I do agree, I think scifi drugs are also excellent, and I definitely had some ideas working on this.

To that end, I actually put some love into the numbers here, the sprites, the process. My google history is absolutely nuclear with how much shit I looked up to make this.

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu commented:
> 👍 also can smokable drugs be put in the toilet bong? would be interactive gameplay

My most esteemed spaceman, these drugs can be put into the toilet bong!

massaheartsu committed:
last interact_with

lmao

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

HalfPunch commented:
I like that pr, yet i have a suggestion. Meth reaction have mechanic of changing meth color acording to it's purity(white->sky blue). I'd like to see crystals reflect that behaviour too.

HalfPunch commented:
Also as far as i can understand your meth crystal reaction produces new object with it's own reagents, and not transfer existing ones. Woudn't that create pure meth? I belive it 100% would purify meth to it's "unreacted purity" value

HalfPunch commented:
> I wanted to avoid 50u morphine plants.
That's the neat part, we already have a way to get 100u+ morphine plants, recipe isn't that hard, and roundstart vendors sell it in good quanities. And that means we don't actually need to make poppies competent source of drugs. I actually suggested scaling it down and not up. Upper value of ~7-10u on potency 100 seems like a good value for balance and it still makes it worth to use better plants

IzaIsAGoat12 commented:
oh wait already suggested, disregard my forsaken mind

massaheartsu committed:
Very significant changes!

Rebalanced both to 10u, added potency scaling for opium, fixed examines and cleaned up useless vars and stuff.

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu commented:
I very much like the idea, still working on it because I'm not a very good coder.

I DID make the opium scale off potency, fixed meth embedding, cleaned up some shitty vars, and the non-transparent pixel on the pipe! This is WAY too many commits for a small project. .

massaheartsu committed:
Update drug_items.dm

fixed that more. fixed black market. prob more fixes.

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu committed:
linters bazinga

linters bazinga

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
FINISHED

Above 90%, meth crystals now gain an increasingly blue tint. Fixed and polished every bit of this.

massaheartsu commented:
@IzaIsAGoat12 @HalfPunch 

I have upgraded the items to hold 10u each, added potency to poppies, and now, above 90% purity, approaching 100%, the rock will gain an increasingly stronger blue tint! Now I feel like this PR is actually complete. Thank you!

<img width="219" height="201" alt="dreamseeker_N8x9bsN3kS-1" src="https://github.com/user-attachments/assets/3af81a91-d08d-48df-8431-2ecb44bf1d5e" />

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu committed:
Update consumables.dm

fixed stupid error

massaheartsu committed:
Update drugs.dm

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

SyncIt21 commented on code/game/objects/items/cigarettes.dm:
```diff
@@ -434,8 +434,9 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 	// allowing reagents to react after being lit
-	reagents.flags &= ~(NO_REACT)
-	reagents.handle_reactions()
+	if(reagents.has_reagent(/datum/reagent/drug/methamphetamine)) //Allows spacemen to smoke meth without exploding!
```
```suggestion
	if(!reagents.has_reagent(/datum/reagent/drug/methamphetamine)) //Allows spacemen to smoke meth without exploding!
```
Mistake on my part. can you test & confirm if it gives your desired output

massaheartsu replied:
Unfortunately it didn't, but it did prompt me to fix my interact_with and stuff that wasn't working. Now meth no longer explodes (again), and you can put the drugs into the thing. 

I can't use interact_with_atom( nor item_interaction(, so I've left it as attackby(. Neither simply work at all to put the drug items into the pipes (different problems with both). But, this code 100000% works, it has the blue tint, no methsplosions, and proper scaling!

massaheartsu committed:
fixed interact

Unfortunately, interact_with simply doesn't work but attackby does

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu committed:
Merge branch 'master' into narcontent

massaheartsu commented:
Yep! Done. I've cleaned it up and tested it thoroughly. Blue meth for the good stuff, scaling, market population, all 9 yards all done and all works, no methpslosions either.

dragomagol reviewed (commented):
Some grammar oddness

massaheartsu committed:
Update code/modules/reagents/chemistry/reagents/medicine_reagents.dm

Co-authored-by: tattle <66640614+dragomagol@users.noreply.github.com>

massaheartsu committed:
Update code/modules/reagents/chemistry/reagents/drug_reagents.dm

Co-authored-by: tattle <66640614+dragomagol@users.noreply.github.com>

massaheartsu committed:
Update code/game/objects/items/drug_items.dm

Co-authored-by: tattle <66640614+dragomagol@users.noreply.github.com>

massaheartsu committed:
Update code/modules/reagents/chemistry/reagents/drug_reagents.dm

Co-authored-by: tattle <66640614+dragomagol@users.noreply.github.com>

massaheartsu committed:
Fixed a few things at the behest of Tattle and Drago!

Fixed thing back to to_smoke, works and tested.

massaheartsu committed:
Merge branch 'narcontent' of https://github.com/massaheartsu/cozystation into narcontent

massaheartsu commented on code/modules/cargo/markets/market_items/consumables.dm:
```diff
@@ -74,3 +74,69 @@
+
+/datum/market_item/consumable/kronkaine
+	name = "Kronkaine"
+	desc = "An 88 ball, it's rare this is ever on the market!"
```
I fixed this in my next commit, but differently.

tttruancy commented:
Actual fucking travesty soul loss PR
